### ISSUE TITLE
understanding pointers to generated classes

### ISSUE THREAD

**Issue by Kimbsen** [therecipe/qt#357]
I'm trying to use a model in go and a listview in qml.

When you have a subclassed type generated it seems the pointer returned from the generated constructor is not the same as the one used. I'm having trouble explaining this so i made some changes to one of the [examples](https://github.com/therecipe/qt/tree/master/internal/examples/sailfish/listview_variant)  and included the diff in the attached file [diff.txt](https://github.com/therecipe/qt/files/1064144/diff.txt)

The change is basically just logging the pointers to the Personmodel instance from two different places in the code. Once in main and once in the constructor of personmodel

The output from running the modified example is:
```
2017/06/09 15:20:40 model.init [0xc4200161e0] [*main.PersonModel]
2017/06/09 15:20:40 main.model [0xc4200162d0] [*main.PersonModel]
```

The tldr is that the pointer returned from main.go:NewPersonModel(nil) is not the same as 'm' in  in personmodel.go:(m *PersonModel) init()

My expectation was that these two pointers would be the same as in that they would point to the same object. 

What i'm actually trying to do in my own application is set some property in the model ( a function ) that i can call when the model changes. But if i set the function in the object returned in main() the function is nil when called by the model. This is obviously due to these being two different objects, but i dont understand why they are.

**Comment by therecipe** [comment]
Hey

Yeah, that's correct the PersonModel pointers can vary because they are just wrappers around the real c++ pointers. If you compare the `(*PersonModel).Pointer()` instead, you will see that both instances refer to the same c++ pointer.

And I just realized that while I made some necessary changes (for some QML use cases) last week. That I broke your use case, because before those changes the `init` function would have been called in the `NewPersonModel` function just right before returning the object.

(The changes I made basically moved the Go constructor "init" call into a separate callback function, so that you can also use Go constructors for QML instantiable classes.)

Maybe take a look into the `moc.go` file and search for `callbackPersonModel_Constructor`, there you will see that the`*PersonModel` instance for the `init` function is just used as a wrapper around the real c++ pointer and not used afterwards.

This is all needed due the "don't take Go pointers into C land" rule, but there are some possible workarounds. And your timing was perfect, as I already had this issue in my mind, but I didn't fully grasped the consequences until now.

There are some possible solutions and if I'm lucky, one of them is also combinable with another problem I'm seeing right now. (pure Go objects in signals/slots/properties)

I will try to fix this asap, but in the meantime you will need to work around this yourself or maybe use `06a1fd8bdfe892906ce66b3a7af29161dc96e994` instead. There it should still work as you expected.

**Comment by Kimbsen** [comment]
Ah i see. I've already worked around this, but it bothered me that i didn't understand why there were two different pointers. Glad to hear it's not intended to be that way and that i will be able to change my code later so it works as i thought it would. Funny you should mention pure go objects in signals/slots as that was my first attempt to work around the problem of making my function callable from the model as slots seemed to be connected to the 'correct' object. Glad to hear its in the pipeline. Thanks for doing an awesome job on these bindings!

**Comment by therecipe** [comment]
The constructor pointer issue should be fixed now with: https://github.com/therecipe/qt/commit/18470b010beffca4b11d7918656a515d7b45d7d4

The pointer is now stored upon the first creation to be later given back through the `New*` functions.

**Comment by therecipe** [comment]
Pure Go types are also supported now: https://github.com/therecipe/qt/commit/2e6a226ed9ba29b4562330ac4f648d4b67f12e26

**Comment by Kimbsen** [comment]
I didn't get a chance to test this before today, but it is working as i originally expected it to :)

Thanks!
